File: java/src/main/java/benchmark/serializers/CapnProtoSer.java
package benchmark.serializers;

import benchmark.capnp.BenchmarkCapnp;
import benchmark.model.Fixture;
import org.capnproto.ArrayInputStream;
import org.capnproto.ArrayOutputStream;
import org.capnproto.MessageBuilder;
import org.capnproto.MessageReader;
import org.capnproto.Serialize;
import org.capnproto.TextList;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.Channels;
import java.util.ArrayList;
import java.util.List;

/**
 * Official Cap'n Proto Java runtime ({@code org.capnproto:runtime}) plus generated
 * {@link BenchmarkCapnp} from the suite {@code .capnp} schema.
 *
 * @see <a href="https://github.com/capnproto/capnproto-java">capnproto-java</a>
 */
public final class CapnProtoSer implements BenchSerializer {
  private String typeId;
  private boolean batch;
  /** Sized in {@link #prepare} so the timed path does not grow or miss. */
  private int writeBufSize = 64 * 1024;

  @Override
  public String name() {
    return "capnproto";
  }

  @Override
  public String version() {
    return Versions.of(Serialize.class);
  }

  @Override
  public String nativeKind() {
    return "schema";
  }

  @Override
  public String streamMode() {
    return "adapted";
  }

  @Override
  public void prepare(Fixture fx) {
    typeId = fx.name;
    batch = TypeUtil.isList(fx.value);
    writeBufSize = sizedBuffer(fx);
  }

  @Override
  public byte[] serializeBytes(Fixture fx) throws Exception {
    MessageBuilder mb = new MessageBuilder();
    fill(mb, fx.value);
    ArrayOutputStream os = new ArrayOutputStream(ByteBuffer.allocate(writeBufSize));
    Serialize.write(os, mb);
    ByteBuffer bb = os.getWriteBuffer().duplicate();
    bb.flip();
    byte[] out = new byte[bb.remaining()];
    bb.get(out);
    return out;
  }

  @Override
  public Object deserializeBytes(byte[] data) throws Exception {
    MessageReader reader = Serialize.read(new ArrayInputStream(ByteBuffer.wrap(data)));
    return readRoot(reader);
  }

  @Override
  public int serializeStream(Fixture fx, java.io.OutputStream out) throws Exception {
    byte[] raw = serializeBytes(fx);
    out.write(raw);
    return raw.length;
  }

  @Override
  public Object deserializeStream(java.io.InputStream in) throws Exception {
    MessageReader reader = Serialize.read(Channels.newChannel(in));
    return readRoot(reader);
  }

  private int sizedBuffer(Fixture fx) {
    MessageBuilder mb = new MessageBuilder();
    fill(mb, fx.value);
    int size = 64 * 1024;
    while (size <= 16 * 1024 * 1024) {
      try {
        Serialize.write(new ArrayOutputStream(ByteBuffer.allocate(size)), mb);
        return size;
      } catch (IOException e) {
        size *= 2;
      }
    }
    throw new IllegalStateException("capnproto payload exceeds 16 MiB");
  }

  private void fill(MessageBuilder mb, Object value) {
    if (batch) {
      fillBatch(mb, (List<?>) value);
      return;
    }
    switch (typeId) {
      case "message" -> fillMessage(mb.initRoot(BenchmarkCapnp.Message.factory), (benchmark.model.v2.Message) value);
      case "document" -> fillDocument(mb.initRoot(BenchmarkCapnp.Document.factory), (benchmark.model.v2.Document) value);
      case "telemetry" -> fillTelemetry(mb.initRoot(BenchmarkCapnp.Telemetry.factory), (benchmark.model.v2.Telemetry) value);
      case "strings" -> fillStrings(mb.initRoot(BenchmarkCapnp.Strings.factory), (benchmark.model.v2.Strings) value);
      case "event" -> fillEvent(mb.initRoot(BenchmarkCapnp.Event.factory), (benchmark.model.v2.Event) value);
      default -> throw new IllegalArgumentException(typeId);
    }
  }

  private void fillBatch(MessageBuilder mb, List<?> list) {
    switch (typeId) {
      case "message" -> {
        var b = mb.initRoot(BenchmarkCapnp.BatchMessage.factory);
        var items = b.initItems(list.size());
        for (int i = 0; i < list.size(); i++) {
          fillMessage(items.get(i), (benchmark.model.v2.Message) list.get(i));
        }
      }
      case "document" -> {
        var b = mb.initRoot(BenchmarkCapnp.BatchDocument.factory);
        var items = b.initItems(list.size());
        for (int i = 0; i < list.size(); i++) {
          fillDocument(items.get(i), (benchmark.model.v2.Document) list.get(i));
        }
      }
      case "telemetry" -> {
        var b = mb.initRoot(BenchmarkCapnp.BatchTelemetry.factory);
        var items = b.initItems(list.size());
        for (int i = 0; i < list.size(); i++) {
          fillTelemetry(items.get(i), (benchmark.model.v2.Telemetry) list.get(i));
        }
      }
      case "strings" -> {
        var b = mb.initRoot(BenchmarkCapnp.BatchStrings.factory);
        var items = b.initItems(list.size());
        for (int i = 0; i < list.size(); i++) {
          fillStrings(items.get(i), (benchmark.model.v2.Strings) list.get(i));
        }
      }
      case "event" -> {
        var b = mb.initRoot(BenchmarkCapnp.BatchEvent.factory);
        var items = b.initItems(list.size());
        for (int i = 0; i < list.size(); i++) {
          fillEvent(items.get(i), (benchmark.model.v2.Event) list.get(i));
        }
      }
      default -> throw new IllegalArgumentException(typeId);
    }
  }

  private static void fillMessage(BenchmarkCapnp.Message.Builder b, benchmark.model.v2.Message m) {
    b.setFBool(m.fBool);
    b.setFInt32(m.fInt32);
    b.setFInt64(m.fInt64);
    b.setFFloat64(m.fFloat64);
    b.setFString(nz(m.fString));
    b.setFBool2(m.fBool2);
    b.setFInt32B(m.fInt32_2);
    b.setFStringB(nz(m.fString2));
  }

  private static void fillDocument(BenchmarkCapnp.Document.Builder b, benchmark.model.v2.Document d) {
    b.setId(nz(d.id));
    b.setStatus(d.status);
    var meta = d.meta != null ? d.meta : new benchmark.model.v2.Document.DocumentMeta();
    var mb = b.initMeta();
    mb.setRegion(nz(meta.region));
    mb.setVersion(meta.version);
    List<benchmark.model.v2.Document.DocumentItem> items = d.items != null ? d.items : List.of();
    var ib = b.initItems(items.size());
    for (int i = 0; i < items.size(); i++) {
      var it = items.get(i);
      var slot = ib.get(i);
      slot.setSku(nz(it.sku));
      slot.setQty(it.qty);
      slot.setPriceMinor(it.priceMinor);
    }
  }

  private static void fillTelemetry(BenchmarkCapnp.Telemetry.Builder b, benchmark.model.v2.Telemetry t) {
    b.setSource(nz(t.source));
    b.setTs(t.ts);
    List<String> tags = t.tags != null ? t.tags : List.of();
    TextList.Builder tb = b.initTags(tags.size());
    for (int i = 0; i < tags.size(); i++) tb.set(i, new org.capnproto.Text.Reader(nz(tags.get(i))));
    double[] vals = t.values != null ? t.values : new double[0];
    var vb = b.initValues(vals.length);
    for (int i = 0; i < vals.length; i++) vb.set(i, vals[i]);
  }

  private static void fillStrings(BenchmarkCapnp.Strings.Builder b, benchmark.model.v2.Strings s) {
    List<String> items = s.items != null ? s.items : List.of();
    TextList.Builder tb = b.initItems(items.size());
    for (int i = 0; i < items.size(); i++) tb.set(i, new org.capnproto.Text.Reader(nz(items.get(i))));
  }

  private static void fillEvent(BenchmarkCapnp.Event.Builder b, benchmark.model.v2.Event e) {
    b.setEventId(nz(e.eventId));
    b.setEventType(nz(e.eventType));
    b.setOccurredAt(e.occurredAt);
    b.setProducer(nz(e.producer));
    List<benchmark.model.v2.Event.EventAttr> attrs = e.attrs != null ? e.attrs : List.of();
    var ab = b.initAttrs(attrs.size());
    for (int i = 0; i < attrs.size(); i++) {
      var a = attrs.get(i);
      var slot = ab.get(i);
      slot.setKey(nz(a.key));
      slot.setValue(nz(a.value));
    }
  }

  private Object readRoot(MessageReader reader) {
    if (batch) return readBatch(reader);
    return switch (typeId) {
      case "message" -> fromMessage(reader.getRoot(BenchmarkCapnp.Message.factory));
      case "document" -> fromDocument(reader.getRoot(BenchmarkCapnp.Document.factory));
      case "telemetry" -> fromTelemetry(reader.getRoot(BenchmarkCapnp.Telemetry.factory));
      case "strings" -> fromStrings(reader.getRoot(BenchmarkCapnp.Strings.factory));
      case "event" -> fromEvent(reader.getRoot(BenchmarkCapnp.Event.factory));
      default -> throw new IllegalArgumentException(typeId);
    };
  }

  private Object readBatch(MessageReader reader) {
    return switch (typeId) {
      case "message" -> {
        var items = reader.getRoot(BenchmarkCapnp.BatchMessage.factory).getItems();
        List<benchmark.model.v2.Message> out = new ArrayList<>(items.size());
        for (int i = 0; i < items.size(); i++) out.add(fromMessage(items.get(i)));
        yield out;
      }
      case "document" -> {
        var items = reader.getRoot(BenchmarkCapnp.BatchDocument.factory).getItems();
        List<benchmark.model.v2.Document> out = new ArrayList<>(items.size());
        for (int i = 0; i < items.size(); i++) out.add(fromDocument(items.get(i)));
        yield out;
      }
      case "telemetry" -> {
        var items = reader.getRoot(BenchmarkCapnp.BatchTelemetry.factory).getItems();
        List<benchmark.model.v2.Telemetry> out = new ArrayList<>(items.size());
        for (int i = 0; i < items.size(); i++) out.add(fromTelemetry(items.get(i)));
        yield out;
      }
      case "strings" -> {
        var items = reader.getRoot(BenchmarkCapnp.BatchStrings.factory).getItems();
        List<benchmark.model.v2.Strings> out = new ArrayList<>(items.size());
        for (int i = 0; i < items.size(); i++) out.add(fromStrings(items.get(i)));
        yield out;
      }
      case "event" -> {
        var items = reader.getRoot(BenchmarkCapnp.BatchEvent.factory).getItems();
        List<benchmark.model.v2.Event> out = new ArrayList<>(items.size());
        for (int i = 0; i < items.size(); i++) out.add(fromEvent(items.get(i)));
        yield out;
      }
      default -> throw new IllegalArgumentException(typeId);
    };
  }

  private static benchmark.model.v2.Message fromMessage(BenchmarkCapnp.Message.Reader r) {
    return new benchmark.model.v2.Message(
        r.getFBool(),
        r.getFInt32(),
        r.getFInt64(),
        r.getFFloat64(),
        r.getFString().toString(),
        r.getFBool2(),
        r.getFInt32B(),
        r.getFStringB().toString());
  }

  private static benchmark.model.v2.Document fromDocument(BenchmarkCapnp.Document.Reader r) {
    var meta = new benchmark.model.v2.Document.DocumentMeta();
    if (r.hasMeta()) {
      meta.region = r.getMeta().getRegion().toString();
      meta.version = r.getMeta().getVersion();
    }
    var itemsIn = r.getItems();
    List<benchmark.model.v2.Document.DocumentItem> items = new ArrayList<>(itemsIn.size());
    for (int i = 0; i < itemsIn.size(); i++) {
      var it = itemsIn.get(i);
      items.add(
          new benchmark.model.v2.Document.DocumentItem(
              it.getSku().toString(), it.getQty(), it.getPriceMinor()));
    }
    return new benchmark.model.v2.Document(r.getId().toString(), r.getStatus(), meta, items);
  }

  private static benchmark.model.v2.Telemetry fromTelemetry(BenchmarkCapnp.Telemetry.Reader r) {
    var tagsIn = r.getTags();
    List<String> tags = new ArrayList<>(tagsIn.size());
    for (int i = 0; i < tagsIn.size(); i++) tags.add(tagsIn.get(i).toString());
    var valsIn = r.getValues();
    double[] vals = new double[valsIn.size()];
    for (int i = 0; i < vals.length; i++) vals[i] = valsIn.get(i);
    return new benchmark.model.v2.Telemetry(r.getSource().toString(), r.getTs(), tags, vals);
  }

  private static benchmark.model.v2.Strings fromStrings(BenchmarkCapnp.Strings.Reader r) {
    var itemsIn = r.getItems();
    List<String> items = new ArrayList<>(itemsIn.size());
    for (int i = 0; i < itemsIn.size(); i++) items.add(itemsIn.get(i).toString());
    return new benchmark.model.v2.Strings(items);
  }

  private static benchmark.model.v2.Event fromEvent(BenchmarkCapnp.Event.Reader r) {
    var attrsIn = r.getAttrs();
    List<benchmark.model.v2.Event.EventAttr> attrs = new ArrayList<>(attrsIn.size());
    for (int i = 0; i < attrsIn.size(); i++) {
      var a = attrsIn.get(i);
      attrs.add(new benchmark.model.v2.Event.EventAttr(a.getKey().toString(), a.getValue().toString()));
    }
    return new benchmark.model.v2.Event(
        r.getEventId().toString(),
        r.getEventType().toString(),
        r.getOccurredAt(),
        r.getProducer().toString(),
        attrs);
  }

  private static String nz(String s) {
    return s == null ? "" : s;
  }
}

File: kotlin/src/main/kotlin/benchmark/serializers/CapnProtoSer.kt
package benchmark.serializers

import benchmark.capnp.BenchmarkCapnp
import benchmark.model.Fixture
import benchmark.model.v2.Document
import benchmark.model.v2.DocumentItem
import benchmark.model.v2.DocumentMeta
import benchmark.model.v2.Event
import benchmark.model.v2.EventAttr
import benchmark.model.v2.Message
import benchmark.model.v2.Strings
import benchmark.model.v2.Telemetry
import org.capnproto.ArrayInputStream
import org.capnproto.ArrayOutputStream
import org.capnproto.MessageBuilder
import org.capnproto.MessageReader
import org.capnproto.Serialize
import org.capnproto.Text
import org.capnproto.TextList
import java.nio.ByteBuffer
import java.nio.channels.Channels

/**
 * Official Cap'n Proto Java runtime (`org.capnproto:runtime`) from Kotlin,
 * plus generated [BenchmarkCapnp] from the suite `.capnp` schema.
 */
class CapnProtoSer : BenchSerializer {
    private var typeId: String = ""
    private var batch = false
    private var writeBufSize = 64 * 1024

    override fun name() = "capnproto"

    override fun version() = Versions.of(Serialize::class.java)

    override fun streamMode() = "adapted"

    override fun nativeKind() = "schema"

    override fun prepare(fx: Fixture) {
        typeId = fx.name
        batch = TypeUtil.isList(fx.value)
        writeBufSize = sizedBuffer(fx)
    }

    override fun serializeBytes(fx: Fixture): ByteArray {
        val mb = MessageBuilder()
        fill(mb, fx.value)
        val os = ArrayOutputStream(ByteBuffer.allocate(writeBufSize))
        Serialize.write(os, mb)
        val bb = os.writeBuffer.duplicate()
        bb.flip()
        val out = ByteArray(bb.remaining())
        bb.get(out)
        return out
    }

    override fun deserializeBytes(data: ByteArray): Any {
        val reader = Serialize.read(ArrayInputStream(ByteBuffer.wrap(data)))
        return readRoot(reader)
    }

    override fun serializeStream(fx: Fixture, out: java.io.OutputStream): Int {
        val raw = serializeBytes(fx)
        out.write(raw)
        return raw.size
    }

    override fun deserializeStream(input: java.io.InputStream): Any {
        val reader = Serialize.read(Channels.newChannel(input))
        return readRoot(reader)
    }

    private fun sizedBuffer(fx: Fixture): Int {
        val mb = MessageBuilder()
        fill(mb, fx.value)
        var size = 64 * 1024
        while (size <= 16 * 1024 * 1024) {
            try {
                Serialize.write(ArrayOutputStream(ByteBuffer.allocate(size)), mb)
                return size
            } catch (_: Exception) {
                size *= 2
            }
        }
        throw IllegalStateException("capnproto payload exceeds 16 MiB")
    }

    private fun fill(mb: MessageBuilder, value: Any) {
        if (batch) {
            fillBatch(mb, value as List<*>)
            return
        }
        when (typeId) {
            "message" -> fillMessage(mb.initRoot(BenchmarkCapnp.Message.factory), value as Message)
            "document" -> fillDocument(mb.initRoot(BenchmarkCapnp.Document.factory), value as Document)
            "telemetry" -> fillTelemetry(mb.initRoot(BenchmarkCapnp.Telemetry.factory), value as Telemetry)
            "strings" -> fillStrings(mb.initRoot(BenchmarkCapnp.Strings.factory), value as Strings)
            "event" -> fillEvent(mb.initRoot(BenchmarkCapnp.Event.factory), value as Event)
            else -> throw IllegalArgumentException(typeId)
        }
    }

    private fun fillBatch(mb: MessageBuilder, list: List<*>) {
        when (typeId) {
            "message" -> {
                val items = mb.initRoot(BenchmarkCapnp.BatchMessage.factory).initItems(list.size)
                for (i in list.indices) fillMessage(items[i], list[i] as Message)
            }
            "document" -> {
                val items = mb.initRoot(BenchmarkCapnp.BatchDocument.factory).initItems(list.size)
                for (i in list.indices) fillDocument(items[i], list[i] as Document)
            }
            "telemetry" -> {
                val items = mb.initRoot(BenchmarkCapnp.BatchTelemetry.factory).initItems(list.size)
                for (i in list.indices) fillTelemetry(items[i], list[i] as Telemetry)
            }
            "strings" -> {
                val items = mb.initRoot(BenchmarkCapnp.BatchStrings.factory).initItems(list.size)
                for (i in list.indices) fillStrings(items[i], list[i] as Strings)
            }
            "event" -> {
                val items = mb.initRoot(BenchmarkCapnp.BatchEvent.factory).initItems(list.size)
                for (i in list.indices) fillEvent(items[i], list[i] as Event)
            }
            else -> throw IllegalArgumentException(typeId)
        }
    }

    private fun fillMessage(b: BenchmarkCapnp.Message.Builder, m: Message) {
        b.fBool = m.fBool
        b.fInt32 = m.fInt32
        b.fInt64 = m.fInt64
        b.fFloat64 = m.fFloat64
        b.setFString(m.fString)
        b.fBool2 = m.fBool2
        b.fInt32B = m.fInt32_2
        b.setFStringB(m.fString2)
    }

    private fun fillDocument(b: BenchmarkCapnp.Document.Builder, d: Document) {
        b.setId(d.id)
        b.status = d.status
        val mb = b.initMeta()
        mb.setRegion(d.meta.region)
        mb.version = d.meta.version
        val ib = b.initItems(d.items.size)
        for (i in d.items.indices) {
            val it = d.items[i]
            val slot = ib[i]
            slot.setSku(it.sku)
            slot.qty = it.qty
            slot.priceMinor = it.priceMinor
        }
    }

    private fun fillTelemetry(b: BenchmarkCapnp.Telemetry.Builder, t: Telemetry) {
        b.setSource(t.source)
        b.ts = t.ts
        val tb: TextList.Builder = b.initTags(t.tags.size)
        for (i in t.tags.indices) tb.set(i, Text.Reader(t.tags[i]))
        val vb = b.initValues(t.values.size)
        for (i in t.values.indices) vb.set(i, t.values[i])
    }

    private fun fillStrings(b: BenchmarkCapnp.Strings.Builder, s: Strings) {
        val tb = b.initItems(s.items.size)
        for (i in s.items.indices) tb.set(i, Text.Reader(s.items[i]))
    }

    private fun fillEvent(b: BenchmarkCapnp.Event.Builder, e: Event) {
        b.setEventId(e.eventId)
        b.setEventType(e.eventType)
        b.occurredAt = e.occurredAt
        b.setProducer(e.producer)
        val ab = b.initAttrs(e.attrs.size)
        for (i in e.attrs.indices) {
            val a = e.attrs[i]
            val slot = ab[i]
            slot.setKey(a.key)
            slot.setValue(a.value)
        }
    }

    private fun readRoot(reader: MessageReader): Any {
        if (batch) return readBatch(reader)
        return when (typeId) {
            "message" -> fromMessage(reader.getRoot(BenchmarkCapnp.Message.factory))
            "document" -> fromDocument(reader.getRoot(BenchmarkCapnp.Document.factory))
            "telemetry" -> fromTelemetry(reader.getRoot(BenchmarkCapnp.Telemetry.factory))
            "strings" -> fromStrings(reader.getRoot(BenchmarkCapnp.Strings.factory))
            "event" -> fromEvent(reader.getRoot(BenchmarkCapnp.Event.factory))
            else -> throw IllegalArgumentException(typeId)
        }
    }

    private fun readBatch(reader: MessageReader): Any =
        when (typeId) {
            "message" -> {
                val items = reader.getRoot(BenchmarkCapnp.BatchMessage.factory).items
                MutableList(items.size()) { fromMessage(items[it]) }
            }
            "document" -> {
                val items = reader.getRoot(BenchmarkCapnp.BatchDocument.factory).items
                MutableList(items.size()) { fromDocument(items[it]) }
            }
            "telemetry" -> {
                val items = reader.getRoot(BenchmarkCapnp.BatchTelemetry.factory).items
                MutableList(items.size()) { fromTelemetry(items[it]) }
            }
            "strings" -> {
                val items = reader.getRoot(BenchmarkCapnp.BatchStrings.factory).items
                MutableList(items.size()) { fromStrings(items[it]) }
            }
            "event" -> {
                val items = reader.getRoot(BenchmarkCapnp.BatchEvent.factory).items
                MutableList(items.size()) { fromEvent(items[it]) }
            }
            else -> throw IllegalArgumentException(typeId)
        }

    private fun fromMessage(r: BenchmarkCapnp.Message.Reader) =
        Message(r.fBool, r.fInt32, r.fInt64, r.fFloat64, r.fString.toString(), r.fBool2, r.fInt32B, r.fStringB.toString())

    private fun fromDocument(r: BenchmarkCapnp.Document.Reader): Document {
        val meta =
            if (r.hasMeta()) DocumentMeta(r.meta.region.toString(), r.meta.version) else DocumentMeta()
        val itemsIn = r.items
        val items =
            MutableList(itemsIn.size()) { i ->
                val it = itemsIn[i]
                DocumentItem(it.sku.toString(), it.qty, it.priceMinor)
            }
        return Document(r.id.toString(), r.status, meta, items)
    }

    private fun fromTelemetry(r: BenchmarkCapnp.Telemetry.Reader): Telemetry {
        val tagsIn = r.tags
        val tags = MutableList(tagsIn.size()) { tagsIn[it].toString() }
        val valsIn = r.values
        val vals = MutableList(valsIn.size()) { valsIn[it] }
        return Telemetry(r.source.toString(), r.ts, tags, vals)
    }

    private fun fromStrings(r: BenchmarkCapnp.Strings.Reader): Strings {
        val itemsIn = r.items
        return Strings(MutableList(itemsIn.size()) { itemsIn[it].toString() })
    }

    private fun fromEvent(r: BenchmarkCapnp.Event.Reader): Event {
        val attrsIn = r.attrs
        val attrs =
            MutableList(attrsIn.size()) { i ->
                val a = attrsIn[i]
                EventAttr(a.key.toString(), a.value.toString())
            }
        return Event(r.eventId.toString(), r.eventType.toString(), r.occurredAt, r.producer.toString(), attrs)
    }
}
